### PR TITLE
Add workflow to deploy the sample index

### PR DESCRIPTION
**Pull request by johnpryan**
This sets up CI to verify and deploy the demos in the web/ directory. It doesn't use `package:peanut`, but I decided to keep the peanut code in case we need to deploy manually for some reason.

The `peaceiris/actions-gh-pages@v3` action keeps the `gh-pages` branch history.

You can see this working on my fork here:
- https://github.com/johnpryan/samples/tree/gh-pages
- http://www.jpryan.me/samples/#
- Verify run: https://github.com/johnpryan/samples/runs/2503174175
- Deploy run: https://github.com/johnpryan/samples/runs/2503174305

### CODE REVIEW

**Comment by johnpryan** [comment]
The verification action is failing due to https://github.com/kevmoo/slide_puzzle/pull/10

File: .github/workflows/gh-pages.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+      - name: Build
+        run: dart _tool/build_ci.dart
+        working-directory: web
+
+      - name: Deploy
+        uses: peaceiris/actions-gh-pages@v3
+        with:
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+          publish_dir: web/samples_index/public
```
Comment (RedBrogdon) [inline .github/workflows/gh-pages.yml:31]: nit: newline

File: web/_tool/common.dart [hunk 2]
```diff
@@ -36,3 +36,16 @@ Future<bool> run(
 void logWrapped(int code, String message) {
   print('\x1B[${code}m$message\x1B[0m');
 }
+
+Iterable<String> listPackageDirs(Directory dir) sync* {
```
Comment (RedBrogdon) [inline web/_tool/common.dart:40]: In all seriousness, how happy were you to find a spot to use a synchronous generator? 😄

**Comment by RedBrogdon** [comment]
I created #793 to fix the underlying broken test. Rebasing on that once it's landed should correct any issues that aren't related to the changes here.

File: web/_tool/common.dart (hunk 2 above)
Comment (johnpryan) [reply web/_tool/common.dart:40]: so stoked.

File: web/_tool/build_ci.dart [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import 'dart:io';
```
Comment (RedBrogdon) [inline web/_tool/build_ci.dart:1]: nit: copyright header

**Review by RedBrogdon — APPROVED** [review]

**Pull request merged**